source/module.c

```diff
@@ -947,10 +947,10 @@ napi_status aws_napi_dispatch_threadsafe_function(
         call_status = status;
         s_handle_failed_callback(env, function, status);
     });
+    /* main thread can exit now */
+    napi_unref_threadsafe_function(env, tsfn);
     /* Must always decrement the ref count, or the function will be pinned */
     napi_status release_status = napi_release_threadsafe_function(tsfn, napi_tsfn_release);
-    /* main thread can exit now */
-    AWS_NAPI_ENSURE(env, napi_unref_threadsafe_function(env, tsfn));
     return (call_status != napi_ok) ? call_status : release_status;
```

Comment:
Theoretical Fix: move unref() call to happen before the release(). [the docs say](https://nodejs.org/api/n-api.html#reference-counting-of-thread-safe-functions):
> napi_release_threadsafe_function() should be the last API call made in conjunction with a given napi_threadsafe_function, because after the call completes, there is no guarantee that the napi_threadsafe_function is still allocated